### DISCUSSION TITLE
Internals demo: e-chirp

### DISCUSSION THREAD

**Discussion by OverLordGoldDragon** [kymatio/kymatio#709] [Q&A] [answered]
Made these as I was stepping through debugger, no coherent code but could make it. I'd appreciate a correctness review; anything looks off?

I'll see if I have time to redo this with larger `N` so frequency rows aren't duplicates. -- FOTS = First-order time scattering, FS = Frequency scattering, SOTS = Second-order. (JFTS -> JTFS, typo)

![image](https://user-images.githubusercontent.com/16495490/113460248-0d248800-9429-11eb-8daa-d04e8fafe387.png)

**Comment by lostanlen** [comment, marked as answer]
impressive work. A few quick comments below

are you sure that this is for an exponential chirp, not linear-frequency (quadratic in phase)? The spectrotemporal pattern in FOTS doesn't look like a straight diagonal line, as i would expect.
Besides, the renormalization factor per octave seems off, otherwise it would show a more even color distribution along the chirp.

> Note, these are output as first-order JTFS, but not used in computing joint variations with 2D wavelets

You're absolutely right.

> Instead, we do joint variability calcs over raw but subsampled first-order coeffs

Yes

> We skip all blue wavelets, so first=orange

Yes

> Pad along freqs (append zero row)

Yes

> These are applied over the freq axis. 

There's a mistake here. It looks like you're reusing the first-order wavelet filterbank (psi) for frequential scattering. But that filterbank has too many wavelets per octave (Q=18), whereas in practice Q=1 or Q=2 is best for the frequency dimension (according to classification experiments but also to evidence from auditory neurophysiology). In other words, we need a custom-built filter bank with a dedicated number of filters per octave Q_fr, defaulting to 1.

Besides, the filterbank for FS must cover the entire spectrum, including negative frequencies (flip the spin) and the zero frequency (include the phi). But you're already aware of that.



The rest looks correct but i can't say much more until the FS filterbank is in order.

**Reply by OverLordGoldDragon** [reply]
Thanks for the feedback.

> There's a mistake here

Oh? It's what we currently have in code; you did set `Q_fr = 1` but then passed in `Q=Q` to `sc_freq=` constructor anyway. I did plan to ask on that - I'll take a look. 

> are you sure that this is for an exponential chirp

You're right, might recall incorrectly

> the renormalization factor per octave seems off

Isn't this something the filter design should've taken care of? Think un-evenness has to do with averaging, but either way unsure one can obtain a perfectly color-even scalogram without specifically tailoring the wavelet; here on echirp, `T, J, Q = 4096, 8, 12`.

![image](https://user-images.githubusercontent.com/16495490/113492825-8dabbd00-94eb-11eb-9632-5d2826bead94.png)

**Reply by lostanlen** [reply]
@janden can you please take a look at this and tell me what you think?

**Reply by lostanlen** [reply]
@OverLordGoldDragon do boundary conditions / padding explain what you've found in this discussion or is there something yet to be understood here?

**Comment by OverLordGoldDragon** [comment]
Re:echirp turns out it's wavelet basics: we're simply not matching the chirp rate, the wavelets dilate far more than the echirp in time, and energy gets spread out. Need a more time-localized wavelet; `kymatio` cannot achieve this without making `Q` too small due to redundancy, but `ssqueezepy` can: here's with an extremely time localized Generalized Morse Wavelet:

![image](https://user-images.githubusercontent.com/16495490/115638776-11570d80-a324-11eb-90d0-b06fa20c39e0.png)

I figure lowest freq imperfections are due to discretization error (which might be impossible to nullify with freq-domain designed filters).

While echirp is a nice theoretical test signal for scattering, one should beware of this effect in implementation; all other coefficients build on top of first-order, and these mismatches will carry over. @lostanlen 

P.S. GMW is superior to Morlet for extreme time localization per perfect analyticity, see e.g. [here](https://user-images.githubusercontent.com/16495490/107903903-d9b69880-6f63-11eb-9478-8ead016cf6f8.png).

**Reply by OverLordGoldDragon** [reply]
@lostanlen Confirming, tracing echirp requires highly time-localized (redundant) wavelets - best I got with Kymatio (with forced analyticity):

<img src="https://user-images.githubusercontent.com/16495490/129242986-0afcfd88-4989-4045-b935-4023bd72c4e8.png" width="450">

This closely matches ssqueezepy's best with GMW (see above), which can trace an instantaneous frequency that itself is white noise, all the way up to Nyquist; basically, if it can't do it, nothing can. 

This suggests the echirp itself isn't actually spinned at lowest frequencies, as a form of discretization error - I also confirmed its Hilbert transform to not be sensible there. Granted, the sweep starts at `fmin=N/(10*N)=0.1`, so discretization errors make sense - results are much better with `fmin=1`, so perhaps I should change the default to that. Kymatio CWT left, ssqueezepy SSQ_CWT right:

![image](https://user-images.githubusercontent.com/16495490/129243174-d47d79fc-4229-43ad-93cc-09b78908dd06.png)

First echirp without forced analyticity for reference:

<img src="https://user-images.githubusercontent.com/16495490/129243403-d6f105f1-52f6-4c92-9dd9-5406b0de56ec.png" width="450">
